### PR TITLE
Enabling limited TensorBoard callback use with Theano and CNTK.

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
## Motivations
- TensorBoard is very useful to visualize metrics and losses. This doesn't technically require to use TensorFlow as a backend. 

- Users should be able to use this functionality within the callback even if TensorFlow isn't the backend. It would also enable users to switch more easily between backends (less code to change).

## Design choices
- If the user has given arguments that are not compatible with the backend (let's say write_graph with Theano or CNTK), the callback won't throw an error, it will just give a warning saying that the argument has been overwritten.
- I make use of `excec` and` eval`, which is debatable in terms of code design. I could do without, it can easily be changed if required. The code would be a bit longer though because I didn't find a clean way to iterate over the arguments.

## Tests
- I believe all the tests needed were already created. I removed the decorators saying to skip them if the backend wasn't TensorFlow.

## API changes
- None

Thank you for reviewing.

### CODE REVIEW

File: keras/callbacks.py [hunk 1]
```diff
@@ -655,9 +657,29 @@ def __init__(self, log_dir='./logs',
...
 
+        if K.backend() != 'tensorflow':
+            compatibility_list = [("histogram_freq", "0"),
+                                  ("write_graph", "False"),
+                                  ("write_images", "False"),
+                                  ("embeddings_freq", "0")]
+            # We find the arguments incompatible with our current backend.
+            incompatible_args = []
+            for x in compatibility_list:
+                if eval(x[0] + " != " + x[1]):
```
Comment (Dref360) [inline keras/callbacks.py]: we could just use a dict instead of doing it. Would be cleaner.

**Comment by gabrieldemarmiesse** [comment]
Thanks for the review! Is this what you meant by using a dict?

**Comment by Dref360** [comment]
Don't we already know what features CNTK/Theano can support?

Why not use : 
```python
not_supported = {'tensorflow':[],
                      'cntk':[...],
                      'theano':[...]
}
```

File: keras/callbacks.py [hunk 2]
```diff
@@ -601,6 +601,10 @@ class TensorBoard(Callback):
     tensorboard --logdir=/full_path_to_your_logs
     ```
 
+    When using a backend other than Tensorflow, Tensorboard will still work
```
Comment (fchollet) [inline keras/callbacks.py]: Use `TensorFlow` and `TensorBoard`everywhere (capitalization)

File: keras/callbacks.py [hunk 3]
```diff
@@ -638,12 +642,14 @@ def __init__(self, log_dir='./logs',
...
-                               'with the TensorFlow backend.')
+
         global tf, projector
-        import tensorflow as tf
-        from tensorflow.contrib.tensorboard.plugins import projector
+        try:
+            import tensorflow as tf
+            from tensorflow.contrib.tensorboard.plugins import projector
+        except ImportError:
+            raise ImportError("You need the TensorFlow module installed to use TensorBoard.")
```
Comment (fchollet) [inline keras/callbacks.py]: For consistency, use `'` as the quote character

File: keras/callbacks.py [hunk 4]
```diff
@@ -655,9 +661,29 @@ def __init__(self, log_dir='./logs',
...
+
+            warning_message = ["You're not using TensorFlow as backend. ",
+                               "Certain functions of the TensorBoard callback are going to be disabled.\n"]
+            display_warning = False
+            # We find the arguments incompatible with our current backend.
+            for key, value in compatibility_dict.items():
+                if self.__dict__[key] != value:
+                    warning_message += [key + " was set to " + str(value), ".\n"]
+                    display_warning = True
+            self.__dict__.update(compatibility_dict)
```
Comment (fchollet) [inline keras/callbacks.py]: Don't do this, prefer explicit checks

File: keras/callbacks.py [hunk 5]
```diff
@@ -655,9 +661,29 @@ def __init__(self, log_dir='./logs',
...
+                                  "write_graph": False,
+                                  "write_images": False,
+                                  "embeddings_freq": 0}
+
+            warning_message = ["You're not using TensorFlow as backend. ",
+                               "Certain functions of the TensorBoard callback are going to be disabled.\n"]
+            display_warning = False
+            # We find the arguments incompatible with our current backend.
+            for key, value in compatibility_dict.items():
+                if self.__dict__[key] != value:
```
Comment (fchollet) [inline keras/callbacks.py]: Don't do this, prefer explicit checks

File: keras/callbacks.py [hunk 6]
```diff
@@ -655,9 +661,29 @@ def __init__(self, log_dir='./logs',
...
         self.batch_size = batch_size
 
+        if K.backend() != 'tensorflow':
+            compatibility_dict = {"histogram_freq": 0,
+                                  "write_graph": False,
+                                  "write_images": False,
+                                  "embeddings_freq": 0}
+
+            warning_message = ["You're not using TensorFlow as backend. ",
+                               "Certain functions of the TensorBoard callback are going to be disabled.\n"]
```
Comment (fchollet) [inline keras/callbacks.py]: I don't think the warning message ends up being very clear... Better to raise specific warnings below, one at a time

**Comment by ahundt** [comment]
Any chance this might also help with tensorflow + input tensors?

Specifically via the following modification to the `callbacks` in [mnist_tfrecord.py](https://github.com/keras-team/keras/blob/master/examples/mnist_tfrecord.py):
```python
test_model_input = layers.Input(tensor=x_test_batch)
test_model_output = cnn_layers(test_model_input)
test_model = keras.models.Model(inputs=test_model_input, outputs=test_model_output)

# Pass the target tensor `y_test_batch` to `compile`
# via the `target_tensors` keyword argument:
test_model.compile(optimizer=keras.optimizers.RMSprop(lr=2e-3, decay=1e-5),
                   loss='categorical_crossentropy',
                   metrics=['accuracy'],
                   target_tensors=[y_test_batch])

# Fit the model using data from the TFRecord data tensors.
coord = tf.train.Coordinator()
threads = tf.train.start_queue_runners(sess, coord)

train_model.fit(epochs=epochs,
                steps_per_epoch=steps_per_epoch,
callbacks=[EvaluateInputTensor(test_model, steps=100), TensorBoard()])
```

**Comment by gabrieldemarmiesse** [comment]
Happy to help!
